File: 3. Longest Substring Without Repeating Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if not s:
            return 0
        else:
            head = 0
            tolmax = 0
            char_dict = {}
            for i, char in enumerate(s):
                if char in char_dict and head <= char_dict[char]:
                    head = char_dict[char] + 1
                char_dict[char] = i
                if tolmax < i - head + 1:
                    tolmax = i - head + 1
            return tolmax

File: 3. Longest Substring Without Repeating Characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        int start = 0;
        int end = 0;
        int result = 0;
        HashMap<Character, Integer> map = new HashMap<>();
        while (end < n) {
            char c = s.charAt(end);
            if (map.containsKey(c) && map.get(c) >= start) {
                start = map.get(c)+1;
            }
            map.put(c, end);
            end++;
            result = Math.max(result, end-start);
        }
        return result;
    }
}